### PR TITLE
Fix #16210 - Show error message and update help for `we` ##io

### LINKED ISSUE #16210: wen Does nothing
Hi, 

Running wen in r2 should insert num null bytes at current offset. It does nothing when running the commands. All commands in we do nothing. This is on Linux

### Work environment

| Questions                                            | Answers
|------------------------------------------------------|--------------------
| OS/arch/bits (mandatory)                             | Linux Archlinux x86_64
| File format of the file you reverse (mandatory)      | ELF
| Architecture/bits of the file (mandatory)            | x86/64
| r2 -v full output, **not truncated** (mandatory)     |  radare2 4.4.0-git 23959 @ linux-x86-64 git.4.3.1-39-g247b28731
commit: 247b2873199143f767ecf8892f23ff35bbfbbc9c build: 2020-03-13__09:56:17

### Expected behavior
wen to expand a few bytes so hex code can be inserted without overwriting code
### Actual behavior
No changes in the binary, no increase in file size
### Steps to reproduce the behavior 

````
$ r2 -v
radare2 4.4.0-git 23959 @ linux-x86-64 git.4.3.1-39-g247b28731
commit: 247b2873199143f767ecf8892f23ff35bbfbbc9c build: 2020-03-13__09:56:17
$ r2 ./hi
[0x00400078]> oo+
[0x00400078]> aa
[x] Analyze all flags starting with sym. and entry0 (aa)
[0x00400078]> pdf
            ;-- rip:
/ 45: entry0 ();
|           ; var int64_t var_4h @ rsp+0x4
|       ,=< 0x00400078      eb0e           jmp 0x400088
..
|       |   ; CODE XREF from entry0 @ 0x400078
|       `-> 0x00400088      50             push rax
|           0x00400089      bf01000000     mov edi, 1
|           0x0040008e      488d359d0000.  lea rsi, [0x00400132]       ; "Hello!\n"
|           0x00400095      ba07000000     mov edx, 7
|           0x0040009a      e819000000     call fcn.004000b8
|           0x0040009f      31ff           xor edi, edi
|           0x004000a1      89442404       mov dword [var_4h], eax
|           0x004000a5      e85e000000     call fcn.00400108
|           0x004000aa      31c9           xor ecx, ecx
|           0x004000ac      890424         mov dword [rsp], eax
|           0x004000af      89c8           mov eax, ecx
|           0x004000b1      59             pop rcx
\           0x004000b2      c3             ret
[0x00400078]> px
- offset -   0 1  2 3  4 5  6 7  8 9  A B  C D  E F  0123456789ABCDEF
0x00400078  eb0e 662e 0f1f 8400 0000 0000 0f1f 4000  ..f...........@.
0x00400088  50bf 0100 0000 488d 359d 0000 00ba 0700  P.....H.5.......
0x00400098  0000 e819 0000 0031 ff89 4424 04e8 5e00  .......1..D$..^.
0x004000a8  0000 31c9 8904 2489 c859 c30f 1f44 0000  ..1...$..Y...D..
0x004000b8  897c 24fc 4889 7424 f089 5424 ec8b 4424  .|$.H.t$..T$..D$
0x004000c8  fc89 4424 dc48 8b4c 24f0 4889 4c24 d08b  ..D$.H.L$.H.L$..
0x004000d8  4424 ec89 4424 cc8b 7c24 dc48 8b74 24d0  D$..D$..|$.H.t$.
0x004000e8  8b54 24cc b801 0000 000f 0548 8944 24e0  .T$........H.D$.
0x004000f8  488b 4424 e089 4424 c88b 4424 c8c3 6690  H.D$..D$..D$..f.
0x00400108  897c 24fc 8b44 24fc 8944 24ec 8b7c 24ec  .|$..D$..D$..|$.
0x00400118  b83c 0000 000f 0548 8944 24f0 488b 4424  .<.....H.D$.H.D$
0x00400128  f089 4424 e88b 4424 e8c3 4865 6c6c 6f21  ..D$..D$..Hello!
0x00400138  0a00 ffff ffff ffff ffff ffff ffff ffff  ................
0x00400148  ffff ffff ffff ffff ffff ffff ffff ffff  ................
0x00400158  ffff ffff ffff ffff ffff ffff ffff ffff  ................
0x00400168  ffff ffff ffff ffff ffff ffff ffff ffff  ................
[0x00400078]> r
490
[0x00400078]> r+10
[0x00400078]> r
500
[0x00400078]> 0x0040009f
[0x0040009f]> wen 10
[0x0040009f]> r
500
[0x0040009f]> pdf
            ;-- rip:
/ 45: entry0 ();
|           ; var int64_t var_4h @ rsp+0x4
|       ,=< 0x00400078      eb0e           jmp 0x400088
..
|       |   ; CODE XREF from entry0 @ 0x400078
|       `-> 0x00400088      50             push rax
|           0x00400089      bf01000000     mov edi, 1
|           0x0040008e      488d359d0000.  lea rsi, [0x00400132]       ; "Hello!\n"
|           0x00400095      ba07000000     mov edx, 7
|           0x0040009a      e819000000     call fcn.004000b8
|           0x0040009f      31ff           xor edi, edi
|           0x004000a1      89442404       mov dword [var_4h], eax
|           0x004000a5      e85e000000     call fcn.00400108
|           0x004000aa      31c9           xor ecx, ecx
|           0x004000ac      890424         mov dword [rsp], eax
|           0x004000af      89c8           mov eax, ecx
|           0x004000b1      59             pop rcx
\           0x004000b2      c3             ret
[0x0040009f]> px
- offset -   0 1  2 3  4 5  6 7  8 9  A B  C D  E F  0123456789ABCDEF
0x0040009f  31ff 8944 2404 e85e 0000 0031 c989 0424  1..D$..^...1...$
0x004000af  89c8 59c3 0f1f 4400 0089 7c24 fc48 8974  ..Y...D...|$.H.t
0x004000bf  24f0 8954 24ec 8b44 24fc 8944 24dc 488b  $..T$..D$..D$.H.
0x004000cf  4c24 f048 894c 24d0 8b44 24ec 8944 24cc  L$.H.L$..D$..D$.
0x004000df  8b7c 24dc 488b 7424 d08b 5424 ccb8 0100  .|$.H.t$..T$....
0x004000ef  0000 0f05 4889 4424 e048 8b44 24e0 8944  ....H.D$.H.D$..D
0x004000ff  24c8 8b44 24c8 c366 9089 7c24 fc8b 4424  $..D$..f..|$..D$
0x0040010f  fc89 4424 ec8b 7c24 ecb8 3c00 0000 0f05  ..D$..|$..<.....
0x0040011f  4889 4424 f048 8b44 24f0 8944 24e8 8b44  H.D$.H.D$..D$..D
0x0040012f  24e8 c348 656c 6c6f 210a 00ff ffff ffff  $..Hello!.......
0x0040013f  ffff ffff ffff ffff ffff ffff ffff ffff  ................
0x0040014f  ffff ffff ffff ffff ffff ffff ffff ffff  ................
0x0040015f  ffff ffff ffff ffff ffff ffff ffff ffff  ................
0x0040016f  ffff ffff ffff ffff ffff ffff ffff ffff  ................
0x0040017f  ffff ffff ffff ffff ffff ffff ffff ffff  ................
0x0040018f  ffff ffff ffff ffff ffff ffff ffff ffff  ................
````

[hi.zip](https://github.com/radareorg/radare2/files/4332396/hi.zip)

### PR DESCRIPTION
**Pull request by radare**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

just read the

**Test plan**

modifying in disk files is a bit tricky and boring, not going to add tests now for this we have more unit tests to add before this imho

**Closing issues**

the #16210  one

### CODE REVIEW

File: libr/core/cmd.c [hunk 1]
```diff
@@ -2884,7 +2884,7 @@ static int r_core_cmd_subst_i(RCore *core, char *cmd, char *colon, bool *tmpseek
 			line = strdup (cmd);
 			line = r_str_replace (line, "\\\"", "\"", true);
 			if (p && *p && p[1] == '|') {
-				str = r_str_trim_head_ro (p + 2);
+				str = (char *)r_str_trim_head_ro (p + 2);
```
Comment (ret2libc) [inline libr/core/cmd.c:2887]: mmm i think you already did this change in another PR.

**Comment by ret2libc** [comment]
> modifying in disk files is a bit tricky and boring, not going to add tests now for this we have more unit tests to add before this imho

I would add a test for this as well. It does not work.

**Comment by radare** [comment]
the wen command works fine:
```
$ echo hello world > txt
$ r2 -w txt
 -- I did it for the pwnz.
[0x00000000]> x 32
- offset -  0001 0203 0405 0607 0809 0A0B 0C0D 0E0F  0123456789ABCDEF
0x00000000  6865 6c6c 6f20 776f 726c 640a ffff ffff  hello world.....
0x00000010  ffff ffff ffff ffff ffff ffff ffff ffff  ................
[0x00000000]> s 3
[0x00000003]> wen 3
[0x00000003]> s 0
[0x00000000]> x 32
- offset -  0001 0203 0405 0607 0809 0A0B 0C0D 0E0F  0123456789ABCDEF
0x00000000  6865 6c00 0000 6c6f 2077 6f72 6c64 0aff  hel...lo world..
0x00000010  ffff ffff ffff ffff ffff ffff ffff ffff  ................
[0x00000000]> r
15
[0x00000000]> q
$ ls -l txt
-rw-r--r--  1 pancake  staff  15 Apr  7 01:21 txt
$ cat txt
hello world
$ hexdump -C txt|head -n 4
00000000  68 65 6c 00 00 00 6c 6f  20 77 6f 72 6c 64 0a     |hel...lo world.|
0000000f
$
```

**Comment by radare** [comment]
only works when io.va=false

**Comment by radare** [comment]
I can improve the help on that to specify that this only makes sense when the underlying io plugin can be resized and it's in PA mode

**Comment by ret2libc** [comment]
> only works when io.va=false
> I can improve the help on that to specify that this only makes sense when the underlying io plugin can be resized and it's in PA mode

That is ok, but please remove the issue #16210 from the title/commits/description, because this is not a *fix* for that issue. Then, if you think #16210 is notabug, close it by explaining your point of view.

If for now `wen` works only when io.va=false, it should be definitely explained in the help msg.

File: libr/core/cmd.c (hunk 1 above)
Comment (radare) [reply libr/core/cmd.c:2887]: yep i think so too. but its not, otherwise i would get a conflict

**Comment by radare** [comment]
```
$ r2 /tmp/ls
 -- This binary no good. Try another.
[0x10000102c]> x 32
- offset -   2C2D 2E2F 3031 3233 3435 3637 3839 3A3B  CDEF0123456789AB
0x10000102c  5548 89e5 4157 4156 4155 4154 5348 81ec  UH..AWAVAUATSH..
0x10000103c  1806 0000 4989 f741 89fe 488d 85c0 fdff  ....I..A..H.....
[0x10000102c]> wen 3
r_io_extend failed
[0x10000102c]> e io.va
true
[0x10000102c]> oo+
[0x10000102c]> x 3
- offset -   2C2D 2E2F 3031 3233 3435 3637 3839 3A3B  CDEF0123456789AB
0x10000102c  5548 89                                  UH.
[0x10000102c]> wen 3
works
[0x10000102c]> x
- offset -   2C2D 2E2F 3031 3233 3435 3637 3839 3A3B  CDEF0123456789AB
0x10000102c  0000 0055 4889 e541 5741 5641 5541 5453  ...UH..AWAVAUATS
0x10000103c  4881 ec18 0600 0049 89f7 4189 fe48 8d85  H......I..A..H..
0x10000104c  c0fd ffff 4889 45d0 85ff 7f05 e858 3100  ....H.E......X1.
```

now works better, i need to add more tests probably

File: libr/core/cmd_write.c [hunk 2]
```diff
@@ -879,13 +879,22 @@ static int cmd_write(void *data, const char *input) {
...
-				len = *input ? r_num_math (core->num, input+3) : 0;
+				len = *input ? r_num_math (core->num, input+ 3) : 0;
 				if (len > 0) {
 					const ut64 cur_off = core->offset;
 					cmd_suc = r_core_extend_at (core, core->offset, len);
-					core->offset = cur_off;
-					r_core_block_read (core);
+					if (cmd_suc) {
+						core->offset = cur_off;
+eprintf ("works\n");
```
Comment (ret2libc) [inline libr/core/cmd_write.c:888]: XX

File: libr/core/cio.c [hunk 3]
```diff
@@ -421,19 +421,24 @@ R_API bool r_core_write_at(RCore *core, ut64 addr, const ut8 *buf, int size) {
...
-		ret = r_io_extend_at (core->io, addr, size);
-		if (addr >= core->offset && addr <= core->offset+core->blocksize) {
-			r_core_block_read (core);
+	int io_va = r_config_get_i (core->config, "io.va");
+	if (io_va) {
+		RIOMap *map = r_io_map_get (core->io, core->offset);
+		if (map) {
+			addr = addr - map->itv.addr + map->delta;
 		}
+		r_config_set_i (core->config, "io.va", io_va);
```
Comment (ret2libc) [inline libr/core/cio.c:434]: why do you set `io_va`, considering it was never changed since you get it?

**Comment by radare** [comment]
i need to add a test where i modify a binary in VA mode, @thestr4ng3r how do you recommend to do taht after the last changes in r2r.c ? do i copy the file into tmp inside r2 and then reopen, and rm when done?

**Comment by thestr4ng3r** [comment]
Yes, see https://github.com/radareorg/radare2/commit/a88ec62304d4c57b8adf7896ae9510e7b28ff82a#diff-f67ce03030de4a37ff248a555a9bb36c

**Comment by radare** [comment]
wat? @ret2libc wen works on io.va=true and false now. it just requires a writable and resizable underlying fd

**Comment by ret2libc** [comment]
> wat? @ret2libc wen works on io.va=true and false now. it just requires a writable and resizable underlying fd

:man_facepalming:  right, sorry :D

**Comment by radare** [comment]
added another test

**Comment by radare** [comment]
ill squash+merge when ci ends

**Pull request merged**